Remove NodeFactory

vampcat opened the pull request:
Yet another Rendering Pipeline Refactoring PR.

Changes:
- Removes the NodeFactory class (and the corresponding injection used by it) for the creation of Nodes.
- Remves the redundant getMaterial(String) function from EnableMaterial StateChange (replacing it with AbstractNode::getMaterial(ResourceUrn)), making the getMaterial calls more uniform
- Renames variables in Nodes and StateChanges to be more descriptive.

GooeyHub commented:
Uh oh, something went wrong with the build. Need to check on that

vampcat commented:
@emanuele3d Can you have a look at this PR, and let me know if I should go ahead and complete it?

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/AlphaRejectBlocksNode.java:
```diff
@@ -54,30 +54,21 @@
+
         wireframeStateChange = new SetWireframe(true);
-        renderingDebugConfig = config.getRendering().getDebug();
+        RenderingDebugConfig renderingDebugConfig = config.getRendering().getDebug();
```
1. unless there are other reasons I'd rewrite this as `context.get(Config.class).getRendering().getDebug();` and I'd avoid storing the config.
2. Can you check if by any chance the RenderingConfig class is available straight from the context?

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/AlphaRejectBlocksNode.java:
```diff
@@ -54,30 +54,21 @@
+        RenderingDebugConfig renderingDebugConfig = config.getRendering().getDebug();
         new WireframeTrigger(renderingDebugConfig, this);
 
         playerCamera = worldRenderer.getActiveCamera();
```
If we need the worldRenderer in a node only to get the playerCamera we might want to write:

`playerCamera = context.get(WorldRenderer.class).getActiveCamera();`

and skip the worldRenderer field altogheter. Some nodes might need it for other reasons though.

vampcat replied:
In this case, it is required for a different reason too. Gonna leave this one as it is.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/AmbientOcclusionNode.java:
```diff
@@ -114,7 +104,7 @@ public void process() {
     @Override
     public void update() {
-        ssaoFbo = displayResolutionDependentFBOs.get(SSAO_FBO);
+        FBO ssaoFbo = displayResolutionDependentFBOs.get(SSAO_FBO);
```
We discussed this already, but please remind me: update() is executed, for example, when the user resize the window. Why not to avoid any form of memory allocation in those circumstances, even if it's admittedly small in this case?

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/BackdropNode.java:
```diff
@@ -60,27 +61,17 @@
+        DisplayResolutionDependentFBOs displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFBOs.class);
+        worldRenderer = context.get(WorldRenderer.class);
+
+        Camera playerCamera = worldRenderer.getActiveCamera();
```
Comments above apply to this node too: please check the need to store the Config and the WorldRenderer.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/BlurNode.java:
```diff
@@ -48,30 +49,22 @@
         this.outputFboUrn = outputFboConfig.getName();
 
-        setupConditions();
+        setupConditions(context);
```
I'm not sure about the necessity of this change and in any case it would probably be good for a separate commit.

GooeyHub commented:
Uh oh, something went wrong with the build. Need to check on that

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Uh oh, something went wrong with the build. Need to check on that

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

vampcat commented on engine/src/main/java/org/terasology/rendering/dag/nodes/CopyImageToScreenNode.java:
```diff
@@ -35,7 +35,7 @@
 
 public class CopyImageToScreenNode extends ConditionDependentNode implements FBOManagerSubscriber {
-    private static final ResourceUrn DEFAULT_FRAME_BUFFER_URN = new ResourceUrn("engine:display");
+    private static final ResourceUrn DEFAULT_FBO = new ResourceUrn("engine:display");
```
I specifically want to point this one out, because it might seem like the new name is actually less descriptive. However, all the Nodes (finally!) follow a common naming scheme, and all the ResourceUrns to Buffers are named as NAME_FBO.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

emanuele3d commented:
@vampcat: thank you for restructuring the commits in this PR - so much quicker to go through it: most commits I just breezed through. 

I don't have the chance right now provide specific feedback but I've only noticed the need for a few minor changes. I'll have some time to highlight those issues sometimes this evening. 

Meanwhile: excellent work!

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/OpaqueBlocksNode.java:
```diff
@@ -48,41 +48,31 @@
+
         wireframeStateChange = new SetWireframe(true);
-        renderingDebugConfig = config.getRendering().getDebug();
+        renderingDebugConfig = context.get(Config.class).getRendering().getDebug();
```
I didn't highlight this issue in previous nodes as it isn't overly important but perhaps we might want to decide if to consistently initialize the two rendering configs either by first obtaining the Config object from the context or, as you do here, by having everything in one line. 

I guess on one hand I feel that having a local variable helps with the readability given that `getRendering()` and `getDebug()` are a bit ambiguous as method names go. On the other I like the compactness of the one-line in this case.

I'd say I'd prefer readability over compactness though. What do you think?

vampcat replied:
Wherever possible, I've tried to go for the one-line code as per your previous comment. However, in very few cases (not sure if Context was one of these cases) other lines used the varaible too, so I went for the "get variable in one line, use in other lines" approach to avoid re-fetching.

Also, the renderingDebugConfig is used in a lot of Nodes, and is probably the only one that gets (slightly) confusing. All the other one-line codes are pretty readable. If you want, I can convert this to two lines.. although I believe we can leave this as it is, since `config.getRendering().getDebug()` isn't much clearer either..

emanuele3d replied:
Alright, let's leave it as it is. And we'll eventually use FCs anyway.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/OpaqueBlocksNode.java:
```diff
@@ -48,41 +48,31 @@
-        addDesiredStateChange(new EnableMaterial(CHUNK_SHADER.toString()));
-        chunkShader = getMaterial(CHUNK_SHADER);
+        addDesiredStateChange(new EnableMaterial(CHUNK_MATERIAL));
+        chunkShader = getMaterial(CHUNK_MATERIAL);
```
Given the renaming occurring elsewhere I guess you might want to use `chunkMaterial` here instead.

vampcat replied:
I don't know how I missed this..
*walks away in shame*

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/ShadowMapNode.java:
```diff
@@ -56,50 +56,43 @@
+        backdropProvider = context.get(BackdropProvider.class);
+        Config config = context.get(Config.class);
+        RenderableWorld renderableWorld = context.get(RenderableWorld.class);
+        ShadowMapResolutionDependentFBOs shadowMapResolutionDependentFBOs = context.get(ShadowMapResolutionDependentFBOs.class);
```
Let's reorganize the section above so that there is the least distance between the line where the variables are first assigned and where the variables are used. I.e. the worldRenderer line can be followed by the playerCamera assignment.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/WorldReflectionNode.java:
```diff
@@ -59,50 +59,44 @@
+        addDesiredStateChange(new EnableMaterial(CHUNK_MATERIAL));
 
         // we must get this here because in process we activate/deactivate a specific shader feature.
         // TODO: improve EnableMaterial to take advantage of shader feature bitmasks.
```
Line below: `chunkShader` --> `chunkMaterial`

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/stateChanges/EnableMaterial.java:
```diff
@@ -16,30 +16,32 @@
 
     private RenderPipelineTask task;
-    private String materialName;
+    private ResourceUrn materialUrn;
```
I'm now wondering if we should call this materialId, given that the resource urn uniquely identify a material. What do you think? Could be done in a separate PR though, it's not that critical.

vampcat replied:
I don't particularly have a problem with that, but MaterialId in the context of OpenGL is generally an int (GLint, to be precise). As things stand, i think it's clear enough, but I leave this one upto you.

emanuele3d replied:
Good point. Let's leave it as it is then.

vampcat commented on engine/src/main/java/org/terasology/rendering/dag/nodes/BlurredAmbientOcclusionNode.java:
```diff
@@ -57,31 +56,24 @@
+    public BlurredAmbientOcclusionNode(Context context) {
+        displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFBOs.class);
+        Config config = context.get(Config.class);
+
```
I used the empty line to separate the context retrieval lines from the rest of the code.

emanuele3d replied:
You mean the config? Sure, but isn't the config used only in the next line?

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/HighPassNode.java:
```diff
@@ -47,21 +47,12 @@
-    public void initialise() {
-        RenderingConfig renderingConfig = config.getRendering();
+    public HighPassNode(Context context) {
+        RenderingConfig renderingConfig = context.get(Config.class).getRendering();
```
This line can be moved below, just before the renderingConfig.subscribe line.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/OpaqueObjectsNode.java:
```diff
@@ -41,35 +41,25 @@
-    public void initialise() {
+    public OpaqueObjectsNode(Context context) {
+        componentSystemManager = context.get(ComponentSystemManager.class);
+        worldRenderer = context.get(WorldRenderer.class);
```
This line can go together with the playerCamera line below...

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/OverlaysNode.java:
```diff
@@ -44,36 +44,26 @@
-    public void initialise() {
+    public OverlaysNode(Context context) {
+        componentSystemManager = context.get(ComponentSystemManager.class);
+        worldRenderer = context.get(WorldRenderer.class);
```
This line can go with the playerCamera line below.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/RefractiveReflectiveBlocksNode.java:
```diff
@@ -56,38 +56,36 @@
-    public void initialise() {
+    public RefractiveReflectiveBlocksNode(Context context) {
+        renderQueues = context.get(RenderQueuesHelper.class);
+        worldRenderer = context.get(WorldRenderer.class);
```
This line can go with the playerCamera line below.
The following line can go closer to the block starting with the requiresFBO line.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/ShadowMapNode.java:
```diff
@@ -56,50 +56,40 @@
-        this.renderingConfig = config.getRendering();
-        renderableWorld.setShadowMapCamera(shadowMapCamera);
+    public ShadowMapNode(Context context) {
+        worldRenderer = context.get(WorldRenderer.class);
```
This line can go with the playerCamera line below.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/WorldReflectionNode.java:
```diff
@@ -59,54 +59,48 @@
+    public WorldReflectionNode(Context context) {
+        renderQueues = context.get(RenderQueuesHelper.class);
+        worldRenderer = context.get(WorldRenderer.class);
+
```
Let's move this empty line above the worldRenderer assignement.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/AmbientOcclusionNode.java:
```diff
@@ -57,31 +57,23 @@
+    private FBO ssaoFbo;
+
+    public AmbientOcclusionNode(Context context) {
+        displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFBOs.class);
```
Let's move this line below, just before the requiresFBO line.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/BlurredAmbientOcclusionNode.java:
```diff
@@ -57,31 +56,23 @@
+    private FBO ssaoBlurredFbo;
+
+    public BlurredAmbientOcclusionNode(Context context) {
+        displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFBOs.class);
```
Let's move this line below, just before the requiresFBO line.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/CopyImageToHMDNode.java:
```diff
@@ -69,12 +57,15 @@
-    public void initialise() {
-        renderingConfig = config.getRendering();
+    public CopyImageToHMDNode(Context context) {
+        displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFBOs.class);
```
Let's move this line to just before the update() line.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/CopyImageToScreenNode.java:
```diff
@@ -36,29 +35,26 @@
-    @Override
-    public void initialise() {
+    public CopyImageToScreenNode(Context context) {
+        displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFBOs.class);
```
Let's move this line to just before the first addDesiredStateChange line.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/DeferredPointLightsNode.java:
```diff
@@ -62,30 +62,20 @@
-        playerCamera = worldRenderer.getActiveCamera();
+    public DeferredPointLightsNode(Context context) {
+        entityManager = context.get(EntityManager.class);
+        DisplayResolutionDependentFBOs displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFBOs.class);
```
Let's move this line to just before the two lines using it.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/HighPassNode.java:
```diff
@@ -47,21 +47,12 @@
-        RenderingConfig renderingConfig = config.getRendering();
+    public HighPassNode(Context context) {
+        RenderingConfig renderingConfig = context.get(Config.class).getRendering();
+        DisplayResolutionDependentFBOs displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFBOs.class);
```
Let's move this line to just before the requiresFBO line.

vampcat commented:
I think I've addressed everything, although I'm a little skeptical about the order of things in `CopyImageToHMDNode`. Right now, I've moved the initialisation of leftEyeFbo and rightEyeFbo to before the `if (vrProvider != null)` block, because that block then uses the value of `leftEyeFbo.colorBufferTextureId`.

However, from what I saw, `requiresCondition` does not "assert" the conditions, it just adds the conditions to it's list and performs a (redundant?) check that returns a boolean value which is then internally ignored. Since we're not asserting the conditions, the only real check for VR I see is the `if (vrProvider != null)` block, so maybe we should move the `requiresFBO(LEFT_EYE_FBO, ...)` and `requiresFBO(rightEyeFbo, ...)` lines inside that block?

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

emanuele3d commented on engine/src/main/java/org/terasology/rendering/dag/nodes/CopyImageToScreenNode.java:
```diff
@@ -36,29 +35,25 @@
+        WorldRenderer worldRenderer = context.get(WorldRenderer.class);
+        displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFBOs.class);
+
         requiresCondition(() -> worldRenderer.getCurrentRenderStage() == MONO || worldRenderer.getCurrentRenderStage() == LEFT_EYE);
```
Weird. I'm not sure what I meant here, but you did the right thing.

emanuele3d commented:
@Cervator tower: this is flight 2892, lined up with runway MRG, requesting permission to land.

Cervator commented:
Runway's clear rubber ducky, bring it on in!

However, your right engine appears to be on fire. Might want to get that checked out :D

Specifically try changing the graphic settings while in-game. I got:

```
21:02:02.998 [Chunk-Generator-1] INFO  o.t.w.b.internal.BlockManagerImpl - Registered Block core:RedLeaf with id 56
21:03:02.801 [main] WARN  o.t.r.nui.databinding.BindHelper - Failed to resolve property oculusVrSupport of type class org.terasology.config.RenderingConfig - is the getter or setter missing?
21:03:04.365 [main] ERROR o.t.e.event.internal.EventSystemImpl - Failed to invoke event
java.lang.NullPointerException: null
	at org.terasology.rendering.dag.ConditionDependentNode.propertyChange(ConditionDependentNode.java:62)
	at java.beans.PropertyChangeSupport.fire(PropertyChangeSupport.java:335)
	at java.beans.PropertyChangeSupport.firePropertyChange(PropertyChangeSupport.java:328)
	at java.beans.PropertyChangeSupport.firePropertyChange(PropertyChangeSupport.java:263)
	at java.beans.PropertyChangeSupport.firePropertyChange(PropertyChangeSupport.java:303)
	at org.terasology.config.RenderingConfig.setLightShafts(RenderingConfig.java:416)
	at org.terasology.rendering.nui.layers.mainMenu.videoSettings.Preset$3.apply(Preset.java:67)
	at org.terasology.rendering.nui.layers.mainMenu.videoSettings.PresetBinding.set(PresetBinding.java:56)
	at org.terasology.rendering.nui.layers.mainMenu.videoSettings.PresetBinding.set(PresetBinding.java:23)
	at org.terasology.rendering.nui.widgets.UIDropdown.setSelection(UIDropdown.java:183)
	at org.terasology.rendering.nui.widgets.UIDropdown$ItemListener.onMouseClick(UIDropdown.java:205)
	at org.terasology.rendering.nui.internal.CanvasImpl.processMouseClick(CanvasImpl.java:224)
	at org.terasology.rendering.nui.internal.NUIManagerInternal.mouseButtonEvent(NUIManagerInternal.java:573)
	at org.terasology.rendering.nui.internal.NUIManagerInternalMethodAccess.invoke(Unknown Source)
	at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:506)
	at org.terasology.entitySystem.event.internal.EventSystemImpl.sendConsumableEvent(EventSystemImpl.java:279)
	at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:258)
	at org.terasology.entitySystem.entity.internal.BaseEntityRef.send(BaseEntityRef.java:145)
	at org.terasology.input.InputSystem.sendMouseEvent(InputSystem.java:545)
	at org.terasology.input.InputSystem.processMouseInput(InputSystem.java:332)
	at org.terasology.input.InputSystem.update(InputSystem.java:279)
	at org.terasology.engine.modes.StateIngame.handleInput(StateIngame.java:186)
	at org.terasology.engine.subsystem.lwjgl.LwjglInput.postUpdate(LwjglInput.java:54)
	at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:422)
	at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:368)
	at org.terasology.engine.Terasology.main(Terasology.java:153)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
21:03:06.556 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Creating game snapshot
```

After that I also got it in the main menu. After deleting my `config.cfg` and starting the game back up I didn't get it in the main menu but *did* get it after loading up the game and trying again.

vampcat commented:
Ahh, the game would turn into a mess if you did not find all these annoying little bugs that elude us all :D

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

Cervator commented:
I hardly find them all :-)

We still need more automated testing and to start using the test plans we got out of GCI (plus adding to those too)

Retested and merged!
